## Exploración Inicial de Datos - Dataset Boston Housing

### By:
Rafael Garcia

### Date:
2026-08-20

### Description:

Este notebook corresponde a la **Tarea 2: Exploración inicial de datos** del Proyecto 1 (Dataset Estático) del curso de MLOps.

El objetivo es realizar una exploración general de los datos para verificar los tipos de datos, comprender las características y el esquema de datos, y solucionar problemas básicos relacionados. Se incluye:

- Descripción general de los datos
- Unificación de la representación de valores nulos
- Conversión de datos a su tipo correcto
- Corrección de inconsistencias de escala detectadas en la Tarea 1
- Almacenamiento del dataset limpio en formato `.parquet`

## 📚 Import libraries

In [1]:
import pandas as pd
import numpy as np

## 💾 Load data

In [2]:
df = pd.read_csv("../../data/01_raw/Precios_Casas_Boston.csv")
print(f"Dimensiones del dataset: {df.shape[0]} filas x {df.shape[1]} columnas")
df.head()

Dimensiones del dataset: 448 filas x 15 columnas


,ID,crim,zn,indus,chas,nox,rm,age,dis,rad,tax,ptratio,black,lstat,medv
0,1,0.00632,18.0,2.31,0.0,538.0,6575.0,65.2,4.0900,1.0,296.0,15.3,396.90,4.98,24.0
1,2,0.02731,0.0,7.07,0.0,469.0,6421.0,78.9,4.9671,2.0,242.0,17.8,396.90,9.14,21.6
2,4,0.03237,0.0,2.18,0.0,458.0,6998.0,45.8,6.0622,3.0,222.0,18.7,394.63,2.94,33.4
3,5,0.06905,0.0,2.18,0.0,458.0,7147.0,54.2,6.0622,3.0,222.0,18.7,396.90,5.33,NaN
4,7,0.08829,12.5,7.87,0.0,524.0,6012.0,66.6,5.5605,5.0,311.0,15.2,395.60,12.43,22.9


## 📋 Descripción General de los Datos

El dataset contiene información sobre precios de viviendas en distintas zonas de Boston con 14 atributos y 1 variable objetivo (`medv`). A continuación se realiza una inspección general del esquema de datos.

In [3]:
# Información general del dataset
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 448 entries, 0 to 447
Data columns (total 15 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   ID       448 non-null    int64  
 1   crim     443 non-null    float64
 2   zn       446 non-null    float64
 3   indus    442 non-null    float64
 4   chas     446 non-null    float64
 5   nox      444 non-null    float64
 6   rm       442 non-null    float64
 7   age      444 non-null    float64
 8   dis      442 non-null    float64
 9   rad      444 non-null    float64
 10  tax      443 non-null    float64
 11  ptratio  446 non-null    float64
 12  black    445 non-null    float64
 13  lstat    447 non-null    float64
 14  medv     432 non-null    float64
dtypes: float64(14), int64(1)
memory usage: 52.6 KB


In [4]:
# Estadísticas descriptivas
df.describe().round(4)

,ID,crim,zn,indus,chas,nox,rm,age,dis,rad,tax,ptratio,black,lstat,medv
count,448.0000,443.0000,446.0000,442.0000,446.0000,444.0000,442.0000,444.0000,442.0000,444.0000,443.0000,446.0000,445.0000,447.0000,432.0000
mean,263.9330,75.8070,9.1491,12.0330,0.0561,479.4523,5590.9481,71.2601,352.5209,9.8018,418.3567,18.4009,347.8032,13.1908,21.9880
std,150.8779,894.0672,20.7944,7.0746,0.2303,227.8047,1991.2407,27.6711,1165.0553,8.7985,171.6113,2.2421,99.9181,7.2434,9.2196
min,1.0000,0.0063,0.0000,0.7400,0.0000,0.4000,4.9700,6.0000,1.1296,1.0000,188.0000,12.6000,3.5000,1.7300,5.0000
25%,140.0000,0.0938,0.0000,5.8600,0.0000,433.0000,5710.0000,50.6750,2.0162,4.0000,284.5000,16.8000,368.7400,7.5250,16.1000
50%,265.5000,0.3523,0.0000,10.0100,0.0000,532.0000,6098.5000,82.1500,3.0992,5.0000,384.0000,19.0000,391.2300,12.0300,20.6000
75%,410.2500,4.5489,0.0000,18.1000,0.0000,614.0000,6437.7500,95.0000,5.5000,24.0000,666.0000,20.2000,395.9900,17.2800,24.7000
max,516.0000,15288.0000,100.0000,27.7400,1.0000,871.0000,8725.0000,100.0000,8344.0000,24.0000,711.0000,21.2000,396.9000,37.9700,50.0000


In [5]:
# Dimensiones y tipos de datos
print(f"Filas: {df.shape[0]}")
print(f"Columnas: {df.shape[1]}")
print(f"\nTipos de datos:")
print(df.dtypes.value_counts())

Filas: 448
Columnas: 15

Tipos de datos:
float64    14
int64       1
Name: count, dtype: int64


### Diccionario de datos

| Columna | Descripción | Tipo esperado | Rango esperado |
|---------|-------------|---------------|----------------|
| `ID` | Identificador del registro | int | 1 - 516 |
| `crim` | Tasa de criminalidad per cápita por zona | float | 0 - 89 |
| `zn` | Proporción de terreno residencial zonificado para lotes > 25,000 sq.ft | float | 0 - 100 |
| `indus` | Proporción de acres de negocios no minoristas por zona | float | 0.74 - 27.74 |
| `chas` | Variable dummy del río Charles (1 si limita con el río, 0 en caso contrario) | int (categórica) | 0, 1 |
| `nox` | Concentración de óxidos de nitrógeno (partes por 10 millones) | float | 0.38 - 0.87 |
| `rm` | Número promedio de habitaciones por vivienda | float | 3.5 - 8.8 |
| `age` | Proporción de unidades ocupadas construidas antes de 1940 | float | 0 - 100 |
| `dis` | Distancia ponderada a cinco centros de empleo de Boston | float | 1.1 - 12.1 |
| `rad` | Índice de accesibilidad a autopistas radiales | int (ordinal) | 1 - 24 |
| `tax` | Tasa de impuesto a la propiedad por $10,000 | float | 187 - 711 |
| `ptratio` | Ratio alumnos-profesor por zona | float | 12.6 - 22 |
| `black` | 1000(Bk - 0.63)² donde Bk es la proporción de personas de raza negra | float | 0.32 - 396.9 |
| `lstat` | Porcentaje de población de estatus socioeconómico bajo | float | 1.73 - 38 |
| `medv` | Valor mediano de viviendas ocupadas (en miles de USD) — **Variable objetivo** | float | 5 - 50 |

## 🔍 Unificar la representación de valores nulos

Se verificará si existen diferentes representaciones de valores nulos en el dataset (NaN, cadenas vacías, valores como -999, "NA", "null", etc.) y se unificarán a `NaN` de pandas.

In [6]:
# Conteo de valores nulos actuales (NaN nativos de pandas)
print("=== Valores nulos por columna ===")
null_counts = df.isnull().sum()
null_pct = (df.isnull().sum() / len(df) * 100).round(2)
null_summary = pd.DataFrame({"Nulos": null_counts, "% Nulos": null_pct})
print(null_summary)
print(f"\nTotal de celdas nulas: {df.isnull().sum().sum()}")
print(f"Porcentaje total: {(df.isnull().sum().sum() / df.size * 100):.2f}%")

=== Valores nulos por columna ===
         Nulos  % Nulos
ID           0     0.00
crim         5     1.12
zn           2     0.45
indus        6     1.34
chas         2     0.45
nox          4     0.89
rm           6     1.34
age          4     0.89
dis          6     1.34
rad          4     0.89
tax          5     1.12
ptratio      2     0.45
black        3     0.67
lstat        1     0.22
medv        16     3.57

Total de celdas nulas: 66
Porcentaje total: 0.98%


In [7]:
# Verificar si hay otros valores que representen nulos
# (cadenas vacías, -999, "NA", "null", "N/A", "?", etc.)
POSSIBLE_NULL_VALUES = ["", " ", "NA", "N/A", "na", "n/a", "null", "NULL", "None", "?", "-", "--", "-999", "999"]

print("=== Búsqueda de representaciones alternativas de nulos ===")
found_alt_nulls = False
for col in df.columns:
    for null_val in POSSIBLE_NULL_VALUES:
        count = (df[col].astype(str) == null_val).sum()
        if count > 0:
            print(f"  Columna '{col}': {count} valores con representación '{null_val}'")
            found_alt_nulls = True

if not found_alt_nulls:
    print("  No se encontraron representaciones alternativas de nulos.")
    print("  Todos los valores faltantes ya están representados como NaN.")

=== Búsqueda de representaciones alternativas de nulos ===
  No se encontraron representaciones alternativas de nulos.
  Todos los valores faltantes ya están representados como NaN.


In [8]:
# Verificar valores negativos inesperados que podrían ser nulos codificados
print("=== Valores negativos por columna ===")
for col in df.select_dtypes(include=[np.number]).columns:
    neg_count = (df[col] < 0).sum()
    if neg_count > 0:
        print(f"  {col}: {neg_count} valores negativos")

if all((df[col] >= 0).all() or df[col].isna().any() for col in df.select_dtypes(include=[np.number]).columns):
    print("  No se encontraron valores negativos inesperados.")

=== Valores negativos por columna ===
  No se encontraron valores negativos inesperados.


**Conclusión sobre nulos:** Todos los valores faltantes del dataset están representados de forma uniforme como `NaN` nativo de pandas. No se encontraron representaciones alternativas de nulos que requieran unificación.

## 🔧 Conversión de tipos de datos y corrección de inconsistencias

En la Tarea 1 se detectaron inconsistencias de escala en las columnas `crim`, `nox`, `rm` y `dis`, donde algunos valores están multiplicados por 1000 respecto a los rangos esperados del dataset original de Boston Housing. En esta sección se corrigen esos problemas y se asignan los tipos de datos correctos.

### Paso 1: Corrección de inconsistencias de escala

Las columnas afectadas y sus umbrales de detección son:

| Columna | Rango esperado | Umbral de detección | Valores afectados |
|---------|---------------|--------------------|-----------|
| `crim` | 0 - 89 | > 100 | Valores multiplicados por ×1000 |
| `nox` | 0.38 - 0.87 | > 1 | Valores multiplicados por ×1000 |
| `rm` | 3.5 - 8.8 | > 100 | Valores multiplicados por ×1000 |
| `dis` | 1.1 - 12.1 | > 100 | Valores multiplicados por ×1000 |

In [9]:
# Crear copia para trabajar
df_clean = df.copy()

# Definir las columnas con problemas de escala y sus umbrales
SCALE_CORRECTIONS = {
    "crim": 100,
    "nox": 1,
    "rm": 100,
    "dis": 100,
}
SCALE_FACTOR = 1000

print("=== Corrección de escala ===")
for col, threshold in SCALE_CORRECTIONS.items():
    mask = df_clean[col] > threshold
    affected = mask.sum()
    print(f"\n{col} (umbral > {threshold}):")
    print(f"  Valores afectados: {affected}")
    print(f"  Antes  -> min: {df_clean[col].min():.4f}, max: {df_clean[col].max():.4f}")
    df_clean.loc[mask, col] = df_clean.loc[mask, col] / SCALE_FACTOR
    print(f"  Después -> min: {df_clean[col].min():.4f}, max: {df_clean[col].max():.4f}")

=== Corrección de escala ===

crim (umbral > 100):
  Valores afectados: 5
  Antes  -> min: 0.0063, max: 15288.0000
  Después -> min: 0.0063, max: 73.5341

nox (umbral > 1):
  Valores afectados: 377
  Antes  -> min: 0.4000, max: 871.0000
  Después -> min: 0.3850, max: 0.8710

rm (umbral > 100):
  Valores afectados: 397
  Antes  -> min: 4.9700, max: 8725.0000
  Después -> min: 3.5610, max: 8.7250

dis (umbral > 100):
  Valores afectados: 44
  Antes  -> min: 1.1296, max: 8344.0000
  Después -> min: 1.1296, max: 10.7103


In [10]:
# Verificar que los rangos corregidos son coherentes con el dataset original
EXPECTED_RANGES = {
    "crim": (0, 89),
    "zn": (0, 100),
    "indus": (0, 28),
    "chas": (0, 1),
    "nox": (0.3, 0.9),
    "rm": (3, 9),
    "age": (0, 100),
    "dis": (1, 13),
    "rad": (1, 24),
    "tax": (180, 720),
    "ptratio": (12, 22),
    "black": (0, 400),
    "lstat": (1, 40),
    "medv": (4, 51),
}

print("=== Verificación de rangos post-corrección ===")
for col, (expected_min, expected_max) in EXPECTED_RANGES.items():
    actual_min = df_clean[col].min()
    actual_max = df_clean[col].max()
    status = "✅" if actual_min >= expected_min * 0.5 and actual_max <= expected_max * 1.5 else "⚠️"
    print(f"  {status} {col}: [{actual_min:.2f}, {actual_max:.2f}] (esperado: [{expected_min}, {expected_max}])")

=== Verificación de rangos post-corrección ===
  ✅ crim: [0.01, 73.53] (esperado: [0, 89])
  ✅ zn: [0.00, 100.00] (esperado: [0, 100])
  ✅ indus: [0.74, 27.74] (esperado: [0, 28])
  ✅ chas: [0.00, 1.00] (esperado: [0, 1])
  ✅ nox: [0.39, 0.87] (esperado: [0.3, 0.9])
  ✅ rm: [3.56, 8.72] (esperado: [3, 9])
  ✅ age: [6.00, 100.00] (esperado: [0, 100])
  ✅ dis: [1.13, 10.71] (esperado: [1, 13])
  ✅ rad: [1.00, 24.00] (esperado: [1, 24])
  ✅ tax: [188.00, 711.00] (esperado: [180, 720])
  ✅ ptratio: [12.60, 21.20] (esperado: [12, 22])
  ✅ black: [3.50, 396.90] (esperado: [0, 400])
  ✅ lstat: [1.73, 37.97] (esperado: [1, 40])
  ✅ medv: [5.00, 50.00] (esperado: [4, 51])


### Paso 2: Conversión de tipos de datos

Se asignan los tipos de datos correctos a cada columna según su naturaleza:

- **`ID`**: entero — identificador, no se usa en el modelo.
- **`chas`**: categórica binaria (0/1) — variable dummy del río Charles.
- **`rad`**: entero (ordinal) — índice de accesibilidad.
- **Resto**: float64 — variables numéricas continuas.

In [11]:
print("=== Tipos de datos antes de conversión ===")
print(df_clean.dtypes)

# Convertir 'chas' a tipo categórico
df_clean["chas"] = df_clean["chas"].astype("Int64").astype("category")

# Convertir 'rad' a entero (nullable porque puede tener nulos)
df_clean["rad"] = df_clean["rad"].astype("Int64")

# Mantener 'ID' como entero
df_clean["ID"] = df_clean["ID"].astype("int64")

# Las demás columnas ya son float64, que es correcto
print("\n=== Tipos de datos después de conversión ===")
print(df_clean.dtypes)

=== Tipos de datos antes de conversión ===
ID           int64
crim       float64
zn         float64
indus      float64
chas       float64
nox        float64
rm         float64
age        float64
dis        float64
rad        float64
tax        float64
ptratio    float64
black      float64
lstat      float64
medv       float64
dtype: object

=== Tipos de datos después de conversión ===
ID            int64
crim        float64
zn          float64
indus       float64
chas       category
nox         float64
rm          float64
age         float64
dis         float64
rad           Int64
tax         float64
ptratio     float64
black       float64
lstat       float64
medv        float64
dtype: object


In [12]:
# Verificar que cada columna tiene un tipo uniforme
print("=== Verificación de uniformidad de tipos ===")
for col in df_clean.columns:
    dtype = df_clean[col].dtype
    non_null = df_clean[col].dropna()
    print(f"  {col}: {dtype} — {len(non_null)} valores no nulos")

=== Verificación de uniformidad de tipos ===
  ID: int64 — 448 valores no nulos
  crim: float64 — 443 valores no nulos
  zn: float64 — 446 valores no nulos
  indus: float64 — 442 valores no nulos
  chas: category — 446 valores no nulos
  nox: float64 — 444 valores no nulos
  rm: float64 — 442 valores no nulos
  age: float64 — 444 valores no nulos
  dis: float64 — 442 valores no nulos
  rad: Int64 — 444 valores no nulos
  tax: float64 — 443 valores no nulos
  ptratio: float64 — 446 valores no nulos
  black: float64 — 445 valores no nulos
  lstat: float64 — 447 valores no nulos
  medv: float64 — 432 valores no nulos


### Paso 3: Revisión de duplicados

In [13]:
# Verificar filas duplicadas (excluyendo ID)
cols_sin_id = [col for col in df_clean.columns if col != "ID"]
duplicados = df_clean.duplicated(subset=cols_sin_id, keep=False)
print(f"Filas duplicadas (excluyendo ID): {duplicados.sum()}")

if duplicados.sum() > 0:
    print("\nRegistros duplicados:")
    print(df_clean[duplicados].sort_values(cols_sin_id))

Filas duplicadas (excluyendo ID): 173

Registros duplicados:
      ID      crim    zn  indus chas    nox     rm    age     dis  rad    tax  \
22    40   0.02763  75.0   2.95    0  0.428  6.595   21.8  5.4011    3  252.0   
338  512   0.02763  75.0   2.95    0  0.428  6.595   21.8  5.4011    3  252.0   
2      4   0.03237   0.0   2.18    0  0.458  6.998   45.8  6.0622    3  222.0   
351    4   0.03237   0.0   2.18    0  0.458  6.998   45.8  6.0622    3  222.0   
23    41   0.03359  75.0   2.95    0  0.428  7.024   15.8  5.4011    3  252.0   
..   ...       ...   ...    ...  ...    ...    ...    ...     ...  ...    ...   
390  428  37.66190   0.0  18.10    0  0.679  6.202   78.7  1.8629   24  666.0   
268  415  45.74610   0.0  18.10    0  0.693  4.519  100.0  1.6582   24  666.0   
383  415  45.74610   0.0  18.10    0  0.693  4.519  100.0  1.6582   24  666.0   
271  419  73.53410   0.0  18.10    0  0.679  5.957  100.0  1.8026   24  666.0   
386  419  73.53410   0.0  18.10    0  0.679  5.9

### Paso 4: Resumen del dataset limpio

In [14]:
print("=" * 60)
print("RESUMEN DEL DATASET DESPUÉS DE LA EXPLORACIÓN INICIAL")
print("=" * 60)
print(f"\nDimensiones: {df_clean.shape[0]} filas x {df_clean.shape[1]} columnas")
print(f"Valores nulos totales: {df_clean.isnull().sum().sum()}")
print(f"\nCorrecciones aplicadas:")
print("  1. Escala corregida en: crim, nox, rm, dis (÷1000 donde correspondía)")
print("  2. chas convertida a tipo categórico (0/1)")
print("  3. rad convertida a entero (Int64, nullable)")
print(f"\nValores nulos restantes (a tratar en tareas posteriores):")
null_remaining = df_clean.isnull().sum()
print(null_remaining[null_remaining > 0])

RESUMEN DEL DATASET DESPUÉS DE LA EXPLORACIÓN INICIAL

Dimensiones: 448 filas x 15 columnas
Valores nulos totales: 66

Correcciones aplicadas:
  1. Escala corregida en: crim, nox, rm, dis (÷1000 donde correspondía)
  2. chas convertida a tipo categórico (0/1)
  3. rad convertida a entero (Int64, nullable)

Valores nulos restantes (a tratar en tareas posteriores):
crim        5
zn          2
indus       6
chas        2
nox         4
rm          6
age         4
dis         6
rad         4
tax         5
ptratio     2
black       3
lstat       1
medv       16
dtype: int64


In [15]:
# Estadísticas descriptivas del dataset corregido
df_clean.describe().round(4)

,ID,crim,zn,indus,nox,rm,age,dis,rad,tax,ptratio,black,lstat,medv
count,448.0000,443.0000,446.0000,442.0000,444.0000,442.0000,444.0000,442.0000,444.0,443.0000,446.0000,445.0000,447.0000,432.0000
mean,263.9330,3.8564,9.1491,12.0330,0.5713,6.2330,71.2601,3.4924,9.8018,418.3567,18.4009,347.8032,13.1908,21.9880
std,150.8779,8.2120,20.7944,7.0746,0.1173,0.6834,27.6711,1.8951,8.7985,171.6113,2.2421,99.9181,7.2434,9.2196
min,1.0000,0.0063,0.0000,0.7400,0.3850,3.5610,6.0000,1.1296,1.0,188.0000,12.6000,3.5000,1.7300,5.0000
25%,140.0000,0.0938,0.0000,5.8600,0.4720,5.8750,50.6750,1.9804,4.0,284.5000,16.8000,368.7400,7.5250,16.1000
50%,265.5000,0.3523,0.0000,10.0100,0.5440,6.1520,82.1500,2.7812,5.0,384.0000,19.0000,391.2300,12.0300,20.6000
75%,410.2500,4.1499,0.0000,18.1000,0.6470,6.5380,95.0000,4.5601,24.0,666.0000,20.2000,395.9900,17.2800,24.7000
max,516.0000,73.5341,100.0000,27.7400,0.8710,8.7250,100.0000,10.7103,24.0,711.0000,21.2000,396.9000,37.9700,50.0000


## 💾 Almacenamiento del dataset en formato Parquet

Se almacena el dataset limpio en formato `.parquet` en la carpeta `data/02_intermediate/`. Este formato es más eficiente que CSV en términos de almacenamiento y velocidad de lectura, y preserva los tipos de datos.

In [16]:
# Guardar en formato parquet
output_path = "../../data/02_intermediate/boston_housing_clean.parquet"
df_clean.to_parquet(output_path, index=False, engine="pyarrow")

# Verificar que se guardó correctamente
df_verify = pd.read_parquet(output_path)
print(f"Archivo guardado en: {output_path}")
print(f"Dimensiones verificadas: {df_verify.shape}")
print(f"Tipos preservados: {(df_verify.dtypes == df_clean.dtypes).all()}")

Archivo guardado en: ../../data/02_intermediate/boston_housing_clean.parquet
Dimensiones verificadas: (448, 15)
Tipos preservados: False


## 📊 Analysis of Results and Conclusions

### Resultados

La exploración inicial del dataset de Boston Housing reveló y corrigió los siguientes problemas:

1. **Valores nulos**: se identificaron 66 celdas nulas distribuidas en todas las columnas. La variable objetivo (`medv`) tiene 16 nulos (3.6%). Todos los nulos ya estaban representados como `NaN` de pandas, no se requirió unificación.

2. **Inconsistencias de escala**: se corrigieron 4 columnas (`crim`, `nox`, `rm`, `dis`) donde algunos valores estaban multiplicados por 1000. La corrección se realizó dividiendo por 1000 los valores por encima de un umbral definido para cada columna. Después de la corrección, todos los valores caen dentro de los rangos esperados del dataset original.

3. **Tipos de datos**: se convirtió `chas` a tipo categórico (variable dummy binaria) y `rad` a entero nullable (`Int64`). Las demás columnas numéricas se mantienen como `float64`.

4. **Formato de almacenamiento**: el dataset limpio se guardó en formato `.parquet` que preserva los tipos de datos y es más eficiente.

### Conclusiones

- El principal problema de calidad del dataset eran las inconsistencias de escala, no los valores nulos.
- Los valores nulos no se imputaron en esta etapa ya que su tratamiento depende del análisis exploratorio (Tarea 3) y del feature engineering (Tarea 4).
- El dataset corregido tiene rangos consistentes con la documentación original de Boston Housing.

## 💡 Proposals and Ideas

1. **EDA detallado (Tarea 3)**: realizar análisis univariable, bivariable y multivariable sobre el dataset corregido para entender las distribuciones y relaciones entre variables.

2. **Estrategia de imputación**: en el EDA se debe evaluar si imputar los nulos con media, mediana, KNN, o eliminar los registros. Para `medv` (16 nulos) se recomienda eliminar esos registros ya que no se puede imputar la variable objetivo.

3. **Análisis de outliers**: con los datos en escala correcta, realizar análisis con boxplots e IQR para identificar outliers reales (no causados por el problema de escala).

4. **Variable `black`**: evaluar éticamente si incluirla en el modelo o descartarla en el feature engineering.

5. **Validación de datos**: crear reglas de validación basadas en los rangos esperados para asegurar la calidad en futuras iteraciones.

## 📖 References

- Guía del proyecto (Tarea 2): [joserzapata.github.io/post/ciencia-datos-proyecto-python/2-exploration/](https://joserzapata.github.io/post/ciencia-datos-proyecto-python/2-exploration/)
- Harrison, D. and Rubinfeld, D.L. (1978). *Hedonic housing prices and the demand for clean air*.
- Pandas documentation: [pandas.pydata.org](https://pandas.pydata.org/docs/)
- Apache Parquet format: [parquet.apache.org](https://parquet.apache.org/)